gc/gc-base/src/main/java/org/projectnessie/versioned/persist/gc/DistributedIdentifyUtil.java

```diff
@@ -0,0 +1,309 @@
+    }
+    return (NessieApiV1) builder.fromConfig(configuration::get).build(NessieApiV1.class);
+  }
+
+  /**
+   * Compute the bloom filter per content id by walking all the live references in a distributed way
+   * using spark.
+   *
+   * @param references list of live references to walk.
+   * @param cutOffTimestampPerRef map of cutoff timestamp per reference.
+   * @param defaultCutOffTimestamp default cutoff timestamp to be used.
+   * @return map of {@link ContentBloomFilter} per content-id.
+   */
+  public Map<String, ContentBloomFilter> getLiveContentsBloomFilters(
+      List<Reference> references,
+      Map<String, Instant> cutOffTimestampPerRef,
+      Instant defaultCutOffTimestamp) {
+    List<Map<String, ContentBloomFilter>> bloomFilterMaps =
+        new JavaSparkContext(session.sparkContext())
+            .parallelize(references, getTaskCount(references))
+            .map(
+                reference ->
+                    computeLiveContents(
+                        config, cutOffTimestampPerRef, defaultCutOffTimestamp, reference))
+            .collect();
+    return mergeLiveContentResults(bloomFilterMaps);
+  }
+
+  /**
+   * Gets the live and expired contents per content id by walking all the live and dead references
+   * in a distributed way using spark and checking the contents against the live bloom filter
+   * results.
+   *
+   * @param liveContentsBloomFilterMap live contents bloom filter per content id.
+   * @param references list of all the references to walk (live and dead)
+   * @return {@link IdentifiedResult} object.
+   */
+  public IdentifiedResult getIdentifiedResults(
+      Map<String, ContentBloomFilter> liveContentsBloomFilterMap, List<Reference> references) {
+
+    List<IdentifiedResult> results =
+        new JavaSparkContext(session.sparkContext())
+            .parallelize(references, getTaskCount(references))
+            .map(reference -> computeExpiredContents(config, liveContentsBloomFilterMap, reference))
+            .collect();
+    // TODO: write each task output directly into the spark table.
+    //  Each executor task can return an iterator of spark ROW
+    //  and it can be written directly to the spark table.
+    //  Instead of accumulating/merging each task output to a single in-memory map.
+    IdentifiedResult identifiedResult = new IdentifiedResult();
+    results.forEach(result -> identifiedResult.addNewContentValues(result.getContentValues()));
+    return identifiedResult;
+  }
+
+  protected Map<String, ContentBloomFilter> computeLiveContents(
+      Map<String, String> config,
+      Map<String, Instant> cutOffTimestampPerRef,
+      Instant defaultCutOffTimestamp,
+      Reference reference) {
+    try (NessieApiV1 api = getApi(config)) {
+      Instant cutOffTimestamp =
+          cutOffTimestampPerRef.getOrDefault(reference.getName(), defaultCutOffTimestamp);
+      return walkLiveReference(
+          api,
+          reference,
+          commitMeta ->
```

Comment:
This predicate is missing a very important check:
All _young_ commits _must_ be considered live. So all commits that happened within e.g. the last 2-3 hours are young - that time should be configurable, but a default of 2h feels fine.
This must happen to ensure that GC does not accidentally consider a freshly written commit "expired"

Reply:
Agree. As it is a two step process, there can be new commits when the step two queries all the commits. 
So, it is good to have configurable `nessie.gc.commit.protection.time.in.hours` (default 2 hours, even though step 1 will ideally finish faster).

I think using it in step 2 is enough. Because If I never consider commits for expiry check in step 2, they can never be expired.

Reply:
Handled.